<a href="https://colab.research.google.com/github/robitussin/CCRNFLRL_EXERCISES/blob/main/Exercise2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 2

In [271]:
from math import sqrt, log
import pandas as pd

## Step 1: Enter your data

Add a new row after each pull: Pull number, Machine played ("A", "B", "C"), and Reward earned.

In [272]:
data = pd.DataFrame({
    "Pull": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
             11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
             21, 22, 23, 24, 25, 26, 27, 28, 29, 30,
             31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
             41, 42, 43, 44, 45, 46, 47, 48, 49, 50],          # Pull number
    "Machine": ["A", "B", "C", "C", "C", "C", "A", "B", "B", "B",
                "C", "B", "C", "A", "A", "A", "A", "A", "A", "A",
                "A", "A", "A", "A", "A", "A", "A", "A", "A", "A",
                "A", "A", "A", "A", "A", "A", "A", "A", "A", "A",
                "A", "A", "A", "A", "A", "A", "A", "A", "A", "A"], # Which machine was played
    "Reward": [0, 0, 3, 0, 0, 0, 0, 2, 0, 0,
               0, 0, 0, 2, 38, 64, 49, 27, 39, 24,
               11, 7, 2, 0, 139, 47, 10, 48, 15, 4,
               107, 7, 60, 15, 17, 26, 36, 35, 22, 40,
               18, 4, 0, 0, 0, 0, 5, 5, 0, 2]         # Reward received from that machine
})

# Step 2: Calculate UCB

In [273]:
def calculate_ucb(df):
    results = []
    total_pulls = len(df)
    machines = sorted(df["Machine"].unique())

    for machine in machines:
        machine_data = df[df["Machine"] == machine]
        n_j = len(machine_data)
        avg_reward = machine_data["Reward"].mean()
        exploration_bonus = sqrt((2 * log(total_pulls)) / n_j) if n_j > 0 else float("inf")
        ucb_score = avg_reward + exploration_bonus
        results.append({
            "Machine": machine,
            "Pulls": n_j,
            "Avg Reward": round(avg_reward, 2),
            "Exploration Bonus": round(exploration_bonus, 4),
            "UCB Score": round(ucb_score, 4)
        })
    return pd.DataFrame(results).sort_values(by="UCB Score", ascending=False)


# Step 3: Calculate and display


In [274]:
ucb_table = calculate_ucb(data)
print("📊 UCB Results Table")
print(ucb_table)

📊 UCB Results Table
  Machine  Pulls  Avg Reward  Exploration Bonus  UCB Score
0       A     39       23.72             0.4479    24.1659
1       B      5        0.40             1.2509     1.6509
2       C      6        0.50             1.1419     1.6419


# Step 4: Next machine suggestion

In [275]:
best_machine = ucb_table.iloc[0]["Machine"]
print(f"\n🎯 Recommended next machine to play: {best_machine}")


🎯 Recommended next machine to play: A


# Results

After 50 timesteps, it's concluded that Machine A (Mystical Wizard) is the most profitable machine with an average of 23.73. Compared to the others, Machines B (Treasures of Egypt) had an 0.40 average while Machine C (Flaming Crates) had 0.50 average.